Description: Support building fat EFI binaries
 - add support for building in a build directory (VPATH stuff)
 - rework previous support for ia32 builds on x86_64
 - add support for x86_64 builds on ia32
Author: Julien BLACHE <jblache@debian.org>

Index: refit-0.14/Make.common
===================================================================
--- refit-0.14.orig/Make.common	2010-04-19 18:31:25.622931971 +0200
+++ refit-0.14/Make.common	2010-04-19 18:31:23.763931557 +0200
@@ -8,13 +8,15 @@
 EFILIB          = /usr/lib
 EFICRT0         = /usr/lib
 
-ARCH            = $(shell uname -m | sed s,i[3456789]86,ia32,)
+HOSTARCH        = $(shell uname -m | sed s,i[3456789]86,ia32,)
+ARCH            := $(HOSTARCH)
 OS		= $(shell uname -s)
 CPPFLAGS        = -I$(EFIINC) -I$(EFIINC)/$(ARCH) -I$(EFIINC)/protocol -DCONFIG_$(ARCH)
 
 OPTIMFLAGS      = -O2 -fno-strict-aliasing
 DEBUGFLAGS      = -Wall
-CFLAGS          = $(OPTIMFLAGS) -mcpu=i386 -fpic -fshort-wchar $(DEBUGFLAGS)
+CFLAGS          = $(ARCH3264) $(OPTIMFLAGS) -fpic -fshort-wchar $(DEBUGFLAGS)
+ASFLAGS         = $(ARCH3264)
 LDFLAGS         = -nostdlib -znocombreloc
 
 prefix          = /usr/bin/
@@ -30,6 +32,31 @@
   CFLAGS       += -frename-registers -mfixed-range=f32-f127
 endif
 
+ifeq ($(ARCH),x86_64)
+  CFLAGS += -DEFI_FUNCTION_WRAPPER -mno-red-zone
+  CPPFLAGS += -DEFIX64
+
+  ifeq ($(HOSTARCH),ia32)
+    ARCH3264 = -m64
+
+    GNUEFILIB := $(GNUEFILIB)64
+    EFILIB    := $(EFILIB)64
+    EFICRT0   := $(EFICRT0)64
+  endif
+endif
+
+ifeq ($(ARCH),ia32)
+  CPPFLAGS += -DEFI32
+
+  ifeq ($(HOSTARCH),x86_64)
+    ARCH3264 = -m32
+
+    GNUEFILIB := $(GNUEFILIB)32
+    EFILIB    := $(EFILIB)32
+    EFICRT0   := $(EFICRT0)32
+  endif
+endif
+
 
 CRTOBJS         = $(EFICRT0)/crt0-efi-$(ARCH).o
 
@@ -44,7 +71,7 @@
 endif
 
 LDFLAGS        += -T $(LDSCRIPT) -shared -Bsymbolic -L$(EFILIB) -L$(GNUEFILIB) $(CRTOBJS)
-LIBS            = -lefi -lgnuefi $(shell $(CC) -print-libgcc-file-name)
+LIBS            = -lefi -lgnuefi $(shell $(CC) $(ARCH3264) -print-libgcc-file-name)
 FORMAT          = efi-app-$(ARCH)
 
 
Index: refit-0.14/gptsync/Makefile.gnuefi
===================================================================
--- refit-0.14.orig/gptsync/Makefile.gnuefi	2010-04-19 18:31:25.483932034 +0200
+++ refit-0.14/gptsync/Makefile.gnuefi	2010-04-19 18:31:23.763931557 +0200
@@ -3,7 +3,11 @@
 # Build control file for the gptsync tool
 #
 
-LOCAL_CPPFLAGS  = -I. -I../include -DEFI32
+SRCDIR = .
+
+VPATH = $(SRCDIR)
+
+LOCAL_CPPFLAGS  = -I$(SRCDIR) -I$(SRCDIR)/../include
 LOCAL_LDFLAGS   = 
 LOCAL_LIBS      = 
 
@@ -12,6 +16,6 @@
 
 all: $(TARGET)
 
-include ../Make.common
+include $(SRCDIR)/../Make.common
 
 # EOF
Index: refit-0.14/gptsync/Makefile.unix
===================================================================
--- refit-0.14.orig/gptsync/Makefile.unix	2010-04-19 18:31:25.537931666 +0200
+++ refit-0.14/gptsync/Makefile.unix	2010-04-19 18:31:23.763931557 +0200
@@ -2,6 +2,10 @@
 # Makefile for gptsync on Unix platforms
 #
 
+SRCDIR = .
+
+VPATH = $(SRCDIR)
+
 RM = rm -f
 CC = gcc
 
@@ -11,7 +15,7 @@
 SHOWPART_TARGET = showpart
 SHOWPART_OBJS   = showpart.unix.o lib.unix.o os_unix.showpart.o
 
-CPPFLAGS = -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I../include
+CPPFLAGS = -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I$(SRCDIR)/../include
 CFLAGS   = -Wall
 LDFLAGS  =
 LIBS     =
Index: refit-0.14/libeg/Makefile
===================================================================
--- refit-0.14.orig/libeg/Makefile	2010-04-19 18:31:25.587932151 +0200
+++ refit-0.14/libeg/Makefile	2010-04-19 18:31:23.763931557 +0200
@@ -3,13 +3,17 @@
 # Build control file for the libeg library
 #
 
-LOCAL_CPPFLAGS  = -I.
+SRCDIR = .
+
+VPATH = $(SRCDIR)
+
+LOCAL_CPPFLAGS  = -I$(SRCDIR)
 
 OBJS            = screen.o image.o text.o load_bmp.o load_icns.o
 TARGET          = libeg.a
 
 all: $(TARGET)
 
-include ../Make.common
+include $(SRCDIR)/../Make.common
 
 # EOF
Index: refit-0.14/refit/Makefile
===================================================================
--- refit-0.14.orig/refit/Makefile	2010-04-19 18:31:25.680931776 +0200
+++ refit-0.14/refit/Makefile	2010-04-19 18:31:23.764932024 +0200
@@ -3,8 +3,20 @@
 # Build control file for the rEFIt boot menu
 #
 
-LOCAL_CPPFLAGS  = -I. -I../include -I../libeg
-LOCAL_LDFLAGS   = -L../libeg
+SRCDIR = .
+
+VPATH = $(SRCDIR)
+
+ifeq ($(ARCH),ia32)
+  LIBEG = build32
+endif
+
+ifeq ($(ARCH),x86_64)
+  LIBEG = build64
+endif
+
+LOCAL_CPPFLAGS  = -I$(SRCDIR) -I$(SRCDIR)/../include -I$(SRCDIR)/../libeg
+LOCAL_LDFLAGS   = -L$(SRCDIR)/../libeg/$(LIBEG)
 LOCAL_LIBS      = -leg
 
 OBJS            = main.o config.o menu.o screen.o icns.o lib.o
@@ -12,6 +24,6 @@
 
 all: $(TARGET)
 
-include ../Make.common
+include $(SRCDIR)/../Make.common
 
 # EOF
